tangelo/linq/simulator.py

```diff
@@ -26,515 +26,35 @@
-        """
-
-        n_qubits = int(math.log2(len(statevector)))
-        frequencies = dict()
-        for i, amplitude in enumerate(statevector):
-            frequency = abs(amplitude)**2
-            if (frequency - self.freq_threshold) >= 0.:
-                frequencies[self.__int_to_binstr(i, n_qubits)] = frequency
-
-        # If n_shots, has been specified, then draw that amount of samples from the distribution
-        # and return empirical frequencies instead. Otherwise, return the exact frequencies
-        if not self.n_shots:
-            return frequencies
-        else:
-            xk, pk = [], []
-            for k, v in frequencies.items():
-                xk.append(int(k[::-1], 2))
-                pk.append(frequencies[k])
-            distr = stats.rv_discrete(name='distr', values=(np.array(xk), np.array(pk)))
-
-            # Generate samples from distribution. Cut in chunks to ensure samples fit in memory, gradually accumulate
-            chunk_size = 10**7
-            n_chunks = self.n_shots // chunk_size
-            freqs_shots = Counter()
-
-            for i in range(n_chunks+1):
-                this_chunk = self.n_shots % chunk_size if i == n_chunks else chunk_size
-                samples = distr.rvs(size=this_chunk)
-                freqs_shots += Counter(samples)
-            freqs_shots = {self.__int_to_binstr_lsq(k, n_qubits): v / self.n_shots for k, v in freqs_shots.items()}
-            return freqs_shots
 
-    def __int_to_binstr(self, i, n_qubits):
-        """Convert an integer into a bit string of size n_qubits, in the order
-        specified for the state vector.
-        """
-        bs = bin(i).split('b')[-1]
-        state_binstr = "0" * (n_qubits - len(bs)) + bs
-        return state_binstr[::-1] if (self.statevector_order == "msq_first") else state_binstr
 
-    def __int_to_binstr_lsq(self, i, n_qubits):
-        """Convert an integer into a bit string of size n_qubits, in the
-        least-significant qubit order.
-        """
-        bs = bin(i).split('b')[-1]
-        state_binstr = "0" * (n_qubits - len(bs)) + bs
-        return state_binstr[::-1]
+def Simulator(target=default_simulator, n_shots=None, noise_model=None, **kwargs) -> SimulatorBase:
+    """Return requested target simulator
+
+    Args:
+        target (string or BaseSimulator): String can be "qiskit", "cirq", "qdk" or "qulacs". Can also provide
+            The child class of BaseSimulator.
+        n_shots (int): Number of shots if using a shot-based simulator.
+        noise_model: A noise model object assumed to be in the format
+            expected from the target backend.
+        kwargs: Other arguments that could be passed to a target. Examples are qubits_to_use for a QPU, transpiler optimization
+            level, error mitigation flags etc.
+
+    Returns:
+        BaseSimulator: The initialized target simulator that is a child class of BaseSimulator.
+    """
+    if target is None:
```

Comment:
Not fond of this code. `target` can be 3 different types (None, a string, or.... anything and we'll try to shove it in L58?) and I'm unclear about how errors (e.g unsupported value for `target`) are currently handled in each of these scenarii.